File: examples/simple-upload.py
"""
Simple Galaksio Client Example (Python)

This example shows how to upload a file to Pinata via the Galaksio broker.

NOTE: This is a simplified example for testing WITHOUT payment.
For production, you need to implement the x402 payment flow.
"""

import requests
import json
from datetime import datetime

BROKER_URL = "http://localhost:8080"


def upload_to_pinata(file_url: str, file_name: str) -> dict:
    """Upload a file to Pinata via Galaksio broker"""
    print("🚀 Uploading file to Pinata via Galaksio...")
    print(f"File URL: {file_url}")
    print(f"File Name: {file_name}")

    payload = {
        "taskType": "storage",
        "fileUrl": file_url,
        "provider": "pinata",
        "meta": {
            "name": file_name,
            "timestamp": datetime.utcnow().isoformat()
        }
    }

    try:
        response = requests.post(
            f"{BROKER_URL}/run",
            json=payload,
            headers={"Content-Type": "application/json"}
        )

        if response.status_code == 402:
            print("\n💰 Payment required!")
            print(f"Payment details: {json.dumps(response.json(), indent=2)}")
            payment_info = response.json()
            if "accepts" in payment_info and len(payment_info["accepts"]) > 0:
                accept = payment_info["accepts"][0]
                print("\nTo proceed:")
                print(f"1. Pay {accept['maxAmountRequired']} USDC to {accept['payTo']}")
                print("2. Get transaction hash")
                print("3. Retry with x-payment header containing proof")
            return None

        response.raise_for_status()

        result = response.json()
        print("\n✅ Upload successful!")
        print(f"Job ID: {result.get('jobId')}")
        print(f"Status: {result.get('status')}")
        print(f"Result: {json.dumps(result.get('result'), indent=2)}")

        return result

    except requests.exceptions.HTTPError as e:
        print(f"\n❌ Upload failed: {e}")
        if e.response is not None:
            print(f"Response: {e.response.text}")
        raise


def check_job_status(job_id: str) -> dict:
    """Check the status of a job"""
    print(f"\n🔍 Checking job status for: {job_id}")

    try:
        response = requests.get(f"{BROKER_URL}/status/{job_id}")
        response.raise_for_status()

        status = response.json()
        print(f"Job status: {json.dumps(status, indent=2)}")
        return status

    except requests.exceptions.HTTPError as e:
        print(f"❌ Failed to check status: {e}")
        if e.response is not None:
            print(f"Response: {e.response.text}")
        raise


def main():
    """Example usage"""
    file_url = "https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf"
    file_name = "test-document.pdf"

    try:
        result = upload_to_pinata(file_url, file_name)

        # Optionally check status
        if result and result.get("jobId"):
            import time
            time.sleep(1)  # Wait 1 second
            check_job_status(result["jobId"])

    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()

File: examples/simple-upload.js
/**
 * Simple Galaksio Client Example (Node.js)
 *
 * This example shows how to upload a file to Pinata via the Galaksio broker.
 *
 * NOTE: This is a simplified example for testing WITHOUT payment.
 * For production, you need to implement the x402 payment flow.
 */

const axios = require('axios');

const BROKER_URL = 'http://localhost:8080';

async function uploadToPinata(fileUrl, fileName) {
  console.log('🚀 Uploading file to Pinata via Galaksio...');
  console.log('File URL:', fileUrl);
  console.log('File Name:', fileName);

  try {
    // Step 1: Send task to broker
    const response = await axios.post(`${BROKER_URL}/run`, {
      taskType: 'storage',
      fileUrl: fileUrl,
      provider: 'pinata',
      meta: {
        name: fileName,
        timestamp: new Date().toISOString()
      }
    });

    console.log('\n✅ Upload successful!');
    console.log('Job ID:', response.data.jobId);
    console.log('Status:', response.data.status);
    console.log('Result:', JSON.stringify(response.data.result, null, 2));

    return response.data;

  } catch (error) {
    if (error.response?.status === 402) {
      console.log('\n💰 Payment required!');
      console.log('Payment details:', JSON.stringify(error.response.data, null, 2));
      console.log('\nTo proceed:');
      console.log('1. Pay', error.response.data.accepts[0].maxAmountRequired, 'USDC to', error.response.data.accepts[0].payTo);
      console.log('2. Get transaction hash');
      console.log('3. Retry with x-payment header containing proof');
    } else {
      console.error('\n❌ Upload failed:', error.response?.data || error.message);
    }
    throw error;
  }
}

async function checkJobStatus(jobId) {
  console.log('\n🔍 Checking job status for:', jobId);

  try {
    const response = await axios.get(`${BROKER_URL}/status/${jobId}`);
    console.log('Job status:', JSON.stringify(response.data, null, 2));
    return response.data;
  } catch (error) {
    console.error('❌ Failed to check status:', error.response?.data || error.message);
    throw error;
  }
}

// Example usage
async function main() {
  const fileUrl = 'https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf';
  const fileName = 'test-document.pdf';

  try {
    const result = await uploadToPinata(fileUrl, fileName);

    // Optionally check status
    if (result.jobId) {
      await new Promise(resolve => setTimeout(resolve, 1000)); // Wait 1 second
      await checkJobStatus(result.jobId);
    }
  } catch (error) {
    // Error already logged
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = { uploadToPinata, checkJobStatus };
